### DISCUSSION TITLE
SPI on Arduino Zero?

### DISCUSSION THREAD

**Discussion by mrjgreg** [Bodmer/TFT_eSPI#1655] [Q&A - General] [answered]
I'm trying to find a solution to setting SPI #define's for the M0+ Arduino Zero? They seem to be set in HW within the SPI Lib and are not # definable as such. In the case of the Std Zero the SPI is on the ICSP port (PortB)
The Adafruit GFX versions work OK naturally defaulting to the hw SPI only needing dc, rst, cs -  8,9,10 to be set
The TFT_eSPI  versions don't seem to work in the same way? Or not recognising the SPI hw defaults?
I've tried a few permutations, but so far not working...? 
Any ideas /proof's  for a working solution?

Cheers
MrGreg

**Comment by Bodmer** [comment]
Assuming it is this board, pinout here:
https://docs.arduino.cc/hardware/zero
SCLK = D13
MISO (CIPO) = D12
MOSI (COPI) = D11

**Reply by mrjgreg** [reply]
Thanks for the prompt reply, but...
Unfortunately, their docu is misleading/confusing or just wrong in places :(
See last page here:
https://docs.arduino.cc/static/b1fbc77f1b73dcd9ac379af596e3472e/ABX00003-full-pinout.pdf
The SPI is actually on the ICSP header, not pins 11,12,13
See also
https://forum.arduino.cc/t/how-to-use-the-spi-on-the-zero/453219

As best I understand the SPI is supposedly set up by SPI.h wrt to the Zero and automagically works without SPI pins being defined at all, well for the adafruit GFX at least?
Something to do with "sercom" ?
There ain't no apparent pin definitions - D numbers published as such that I have found so far.
Anyways
Many thanks for looking at the problem, what with it not being one of your supported boards
Cheers & thanks
MrGreg

**Comment by mrjgreg** [comment, marked as answer]
I have found the answer :)
Long read....
Short version
#define TFT_SCLK 24 
#define TFT_MISO 22 
#define TFT_MOSI 23 

Sets the #defines for the existing hw SPI on the ICSP header
So it now works with TFT_eSPI

Pity they didn't publish that in the Zero tech specs

Long version
See:
https://learn.adafruit.com/using-atsamd21-sercom-to-add-more-spi-i2c-serial-ports?view=all

Quite a read, however there are solutions to set the SPI to other available pins Eg 11, 12, 13 and indeed other comms can be re tasked as required if available. Bit of mucking about needed.

Cheers
MrGreg
